gobblin-metrics/src/main/java/gobblin/metrics/MetricReportUtils.java

```diff
@@ -0,0 +1,106 @@
+ */
+
+package gobblin.metrics;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.specific.SpecificDatumReader;
+
+import com.google.common.base.Optional;
+import com.google.common.io.Closer;
+
+
+/**
+ * Utilities for {@link gobblin.metrics.MetricReport}.
+ */
+public class MetricReportUtils {
+
+  public static final int SCHEMA_VERSION = 1;
+  private static Optional<SpecificDatumReader<MetricReport>> READER = Optional.absent();
+
+  /**
+   * Parses a {@link gobblin.metrics.MetricReport} from a byte array representing a json input.
+   * @param reuse MetricReport to reuse.
+   * @param bytes Input bytes.
+   * @return MetricReport.
+   * @throws java.io.IOException
+   */
+  public synchronized static MetricReport deserializeReportFromJson(MetricReport reuse, byte[] bytes) throws IOException {
+    if (!READER.isPresent()) {
+      READER = Optional.of(new SpecificDatumReader<MetricReport>(MetricReport.class));
+    }
+
+    Closer closer = Closer.create();
+
+    try {
+      DataInputStream inputStream = closer.register(new DataInputStream(new ByteArrayInputStream(bytes)));
+
+      // Check version byte
+      int versionNumber = inputStream.readInt();
+      if (versionNumber != SCHEMA_VERSION) {
+        throw new IOException(String
+            .format("MetricReport schema version not recognized. Found version %d, expected %d.", versionNumber,
+                SCHEMA_VERSION));
+      }
+
+      // Decode the rest
+      Decoder decoder = DecoderFactory.get().jsonDecoder(MetricReport.SCHEMA$, inputStream);
+      return READER.get().read(reuse, decoder);
+    } catch(Throwable t) {
+      throw closer.rethrow(t);
+    } finally {
+      closer.close();
+    }
+  }
+
+  /**
+   * Parses a {@link gobblin.metrics.MetricReport} from a byte array Avro serialization.
+   * @param reuse MetricReport to reuse.
+   * @param bytes Input bytes.
+   * @return MetricReport.
+   * @throws java.io.IOException
+   */
+  public synchronized static MetricReport deserializeReportFromAvroSerialization(MetricReport reuse, byte[] bytes)
+      throws IOException {
+    if (!READER.isPresent()) {
+      READER = Optional.of(new SpecificDatumReader<MetricReport>(MetricReport.class));
+    }
+
+    Closer closer = Closer.create();
+
+    try {
+      DataInputStream inputStream = closer.register(new DataInputStream(new ByteArrayInputStream(bytes)));
+
+      // Check version byte
+      int versionNumber = inputStream.readInt();
+      if (versionNumber != SCHEMA_VERSION) {
+        throw new IOException(String
+            .format("MetricReport schema version not recognized. Found version %d, expected %d.", versionNumber,
+                SCHEMA_VERSION));
+      }
+
+      // Decode the rest
+      Decoder decoder = DecoderFactory.get().binaryDecoder(inputStream, null);
```

Comment:
the DataInputStream is used to read the schema version, which is an integer at the start of the byte array (lines 89-94).